### PR TITLE
[PM-1320] Argon2 SIMD support

### PR DESCRIPTION
**Pull request by quexten**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This PR changes the argon2-browser dependency to argon2-browser-bw (a fork I created since argon2-browser does not seem to be under active development at the moment). The fork is different in that it includes wasm binaries compiled using a newer Emscripten compiler version, which leads to significantly faster binaries (exact differences can be found on this [GitHub](https://github.com/antelle/argon2-browser/compare/master...quexten:argon2-browser:bitwarden) branch). Furthermore, it adds automatic detection for SIMD support, and loads the correct WASM module. This makes unlock times on high argon2 configurations a lot more bearable.

On my system the unlock times are as follows (10 iterations, 1 parallelism, 1024MiB memory)
Current implementation in Bitwarden (non-SIMD): 26 seconds
argon2-browser SIMD: 14 seconds
argon2-browser-bw (also SIMD): 10 seconds

The SIMD improvements work on all modern chromium and firefox based browsers and the Electron desktop client. Safari support should arrive in version 16.4.

## Code changes

- webpack.main.js: Add the simd webassembly to be copied to the desktop build
- webCryptoFunction.service.ts: Replace the dependency
- package-lock.json: Replace the dependency, wasm-feature-detect is added as a further package my argon2-browser-bw
- package.json: Replace the dependency

### CODE REVIEW

**Comment by MGibson1** [comment]
Discussed [elsewhere](https://github.com/orgs/bitwarden/discussions/4777#discussioncomment-5256981) as well, but the [upstream solution](https://github.com/antelle/argon2-browser/pull/89) would be preferable.

**Comment by quexten** [comment]
As also discussed in the discussion, the upstream argon2-browser library seems currently unmaintained for the foreseeable future. 

In case that does not change, would you be ok with applying the auto-loading of the SIMD WebAssembly change as a patch? That way the original package is still used, the changes are applied, and can be easily undone once / if upstream is merged. (https://www.npmjs.com/package/patch-package, a solution as recommended f.e here: https://stackoverflow.com/a/62567504).

**Comment by MGibson1** [comment]
I've never seen patch-package before, but I like it a lot. It seems like a reasonable way to solve our issue without implicitly getting us in the business of maintaining an npm package.

**Comment by Hinton** [comment]
@MGibson1 we used patches to handle some issues with electron-builder a long time ago. https://github.com/bitwarden/desktop/pull/803

**Comment by quexten** [comment]
Finally got to revisit this PR. I have re-based to latest master and changed it to just use a small patch to add automatic detection and loading of the SIMD binary instead of the forked package. 

This also means that the WASM binaries are still the same ones from argon2-browser which are somewhat slower than ones newly compiled using the latest emscripten version. But there is not much that can be done about this aside from getting a new release compiled upstream (or adding a postinstall script that copies the binaries).

I also tested the argon2-browser's SIMD implementation on Safari 16.4. It seems to run, but is not actually faster than the regular implementation though this was on arm (A12Z Bionic).

Finally, since https://github.com/bitwarden/clients/pull/5072 also applies a patch to argon2-browser, I will update it (or this PR depending on which is merged first) to apply both changes.

File: patches/argon2-browser+1.18.0.patch [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+         }
+     }
+ 
++    // ref: https://stackoverflow.com/a/47880734/1090359
++    // ref: https://github.com/GoogleChromeLabs/wasm-feature-detect/blob/main/src/detectors/simd/module.wat (translated with wat2wasm)
++    function checkIfSIMDSupported() {
++        try {
++            if (typeof WebAssembly === "object" && typeof WebAssembly.instantiate === "function") {
++                const module = new WebAssembly.Module(
++                    Uint8Array.of(0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00, 0x01, 0x05, 0x01, 0x60, 0x00, 0x01, 0x7b, 0x03, 0x02, 0x01, 0x00, 0x0a, 0x0a, 0x01, 0x08, 0x00, 0x41, 0x00, 0xfd, 0x0f, 0xfd, 0x62, 0x0b)
```
[the remark this answers is not in the document]
Comment (quexten) [reply patches/argon2-browser+1.18.0.patch:49]: No Idea what the difference is. I just ran with version 1.0.32 on Fedora 38 and end up with the same result as before: 
image

Could be due to a different wat2wasm version.

[the remark this answers is not in the document]
Comment (MGibson1) [reply patches/argon2-browser+1.18.0.patch:49]: weird, whatever. Thanks for checking

**Review by MGibson1 — APPROVED** [review]
g2g, Passed QA and just merged up!

Thank you again for your continued support on the argon front, @quexten! I want to send you some swag for all your help here, is the email address you have listed on GitHub a good one to send a giftcard?

**Comment by quexten** [comment]
Thanks @MGibson1 for taking the time for reviewing it, getting it QA'd and merged :) 

No need for a giftcard, I already received one previously from Kyle, but thank you for the offer.

I'll re-base the some-what related https://github.com/bitwarden/clients/pull/5072 onto the current master some time later this week as it was in conflict with this branch.

**Pull request merged**